### PR TITLE
Audio data transfer

### PR DESCRIPTION
**Pull request by paddatrapper**
~This is blocked by PR #23 as it is an extension of that work~. Generates a tone when flashed on an FX2

### CODE REVIEW

**Comment by paddatrapper** [comment]
Travis is failing with the following, seems to be an issue with the `version_data.sh` script. It fails on 

    git checkout $TRAVIS_COMMIT -b $TRAVIS_BRANCH

It seems to be an issue in Travis only, as building locally works fine. I'm not sure the solution.

    make -C audio
    make[1]: Entering directory `/home/travis/build/timvideos/HDMI2USB-fx2-firmware/audio'
         GEN  version_data.h
    remote: Counting objects: 564, done.
    remote: Compressing objects: 100% (187/187), done.
    remote: Total 564 (delta 266), reused 551 (delta 257), pack-reused 0
    Receiving objects: 100% (564/564), 130.73 KiB | 0 bytes/s, done.
    Resolving deltas: 100% (266/266), completed with 14 local objects.
    Fixing detached head
    * (HEAD detached at e0c22de) e0c22de Merge 950615fc869f350033dccd3937895135be0b4826 into c84892ecc756d98e0e7640b0799cf0ffb89c2073
      master                     c84892e Merge pull request #37 from paddatrapper/version-embed
    Deleted branch master (was c84892e).
    fatal: reference is not a tree: 75bebb54743bae46e16e6e7e460f30270429500b
    make[1]: *** [version_data.h] Error 128
    make[1]: Leaving directory `/home/travis/build/timvideos/HDMI2USB-fx2-firmware/audio'
    make: *** [audio/audio.hex] Error 2

**Comment by mithro** [comment]
So, this is a weirdness of Travis the commit you are trying to checkout doesn't exist in the tree.

**Comment by mithro** [comment]
So, https://github.com/timvideos/HDMI2USB-litex-firmware/blob/master/.travis/setup.sh does the following,

```bash
echo ""
echo ""
echo ""
echo "- Fetching non shallow to get git version"
echo "---------------------------------------------"
git fetch --unshallow && git fetch --tags
if [ z"$TRAVIS_BRANCH" != z ]; then
	echo "Fixing detached head"
	git branch -v
	git branch -D $TRAVIS_BRANCH || true
	git checkout $TRAVIS_COMMIT -b $TRAVIS_BRANCH
	git branch -v
fi
GIT_REVISION=`git describe`
echo "============================================="
```

**Comment by paddatrapper** [comment]
We do the same and that checkout is what is failing

https://github.com/timvideos/HDMI2USB-fx2-firmware/blob/master/common/version_data.sh#L11

**Comment by mithro** [comment]
I can't find the commit `75bebb54743bae46e16e6e7e460f30270429500b` anywhere in repos...

**Comment by paddatrapper** [comment]
This is now ready for review and merge

File: audio/Makefile [hunk 1]
```diff
@@ -38,10 +38,10 @@ LINUXHEADERSDIR := ../third_party/fx2lib-linux-headers
 COMMON_DIR := ../common
 INCS := -I$(LINUXHEADERSDIR)
 
-CFLAGS := -DDSCR_AREA=0x3e00 --code-size 0x3c00 --xram-size 0x0200
+CFLAGS := -DDSCR_AREA=0x3d00 --code-size 0x3c00 --xram-size 0x0200
```
Comment (mithro) [inline audio/Makefile]: Why 0x3d00? I assume you need more space for the descriptors?

Maybe add a comment here?

File: audio/Makefile [hunk 2]
```diff
@@ -38,10 +38,10 @@ LINUXHEADERSDIR := ../third_party/fx2lib-linux-headers
...
 CFLAGS += --xram-loc 0x3c00 -Wl"-b INT2JT=0x3f00"
 
-CC_SRCS := descriptors.c firmware.c version_data.c
+CC_SRCS := descriptors.c debug.c audiodata.c firmware.c version_data.c
```
Comment (mithro) [inline audio/Makefile:46]: `debug.c`?

File: audio/TODO [hunk 3]
```diff
@@ -3,3 +3,5 @@
   capture
   http://www.usb.org/developers/docs/whitepapers/iadclasscode_r10.pdf
   http://www.usb.org/developers/docs/InterfaceAssociationDescriptor_ecn.pdf
+* Add more supported frequencies (descriptors.h descripts.highspeed.format)
+* Add interrupts
```
Comment (mithro) [inline audio/TODO]: Is this still accurate?

File: audio/audiodata.c [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+#include <fx2regs.h>
+#include <delay.h>
+#include <eputils.h>
+
+#include "audiodata.h"
+
+#define SYNCDELAY SYNCDELAY4
+
+/**
+ * Returns the configuration. We only support cfg 1
```
Comment (mithro) [inline audio/audiodata.c]: Why do you only support cfg 1?

File: audio/audiodata.c [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    return cfg==1 ? TRUE : FALSE;
+}
+
+/* The current alternative setting */
+BYTE alt_setting = 0;
+/**
+ * Returns the interface currently in use.
+ */
+BOOL handle_get_interface(BYTE ifc, BYTE* alt_ifc) {
+    printf("Get Interface\n");
```
Comment (mithro) [inline audio/audiodata.c:57]: Can you change the printf's to work like the ones here -> https://github.com/mithro/fx2lib/blob/master/lib/setupdat.c#L21-L26
```c
#ifdef DEBUG_SETUPDAT
#include <stdio.h> // NOTE this needs deleted
#else
#define printf(...)
#define NULL (void*)0;
#endif
```

File: audio/audiodata.c [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    return FALSE;
+}
+
+/**
+ * Sets the interface in use.
+ * 0,0 - Default control
+ * 1,0 - Streaming, no endpoint. This is used for when the device is not
+ *       streaming
+ * 1,1 - Streaming with endpoint 4.
+ * SEE TRM 2.3.7
```
Comment (mithro) [inline audio/audiodata.c]: Can you change `SEE TRM 2.3.7` to something which means a person could directly find the info? IE A link to the TRM and is 2.3.7 a section, figure, page number?

File: audio/audiodata.c [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        SYNCDELAY; EP6CFG = 0x7F;
+        SYNCDELAY; EP8CFG = 0x7F;
+        SYNCDELAY; RESETFIFO(0x02);
+        /* reset toggles */
+        SYNCDELAY; RESETTOGGLE(0x82);
+        return TRUE;
+    } else if (ifc == 1 && alt_ifc == 1) {
+        alt_setting = 1;
+        /* Reset audio streaming endpoint */
+        EP8CFG = (bmVALID | bmDIR | bmTYPE0);
```
Comment (mithro) [inline audio/audiodata.c:102]: EP8 or EP4? Your comment mentions something about EP4?

File: audio/debug.c [hunk 8]
```diff
@@ -0,0 +1,170 @@
+// Copyright (C) 2009-2012 Chris McClelland
```
Comment (mithro) [inline audio/debug.c:1]: Should this be part of `common`?

File: audio/debug.h [hunk 9]
```diff
@@ -0,0 +1,36 @@
+// Copyright (C) 2009-2012 Chris McClelland
```
Comment (mithro) [inline audio/debug.h:1]: Should this be part of common?

File: audio/debug.c [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+// You should have received a copy of the GNU Lesser General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+/** \file debug.c
+ * Implements writing to serial for debugging purposes
+ */
+#include <fx2regs.h>
+#include "debug.h"
+
+#define PD3 0xB3
```
Comment (mithro) [inline audio/debug.c:23]: This is only for fx2 mini devboard right?

Comment (mithro) [reply audio/debug.c:23]: `PD3` is TMS on the Atlys board and not sure what it is used for on the Opsis?

File: audio/descriptors.c [hunk 11]
```diff
@@ -26,10 +28,13 @@ __code __at(DSCR_AREA) struct usb_descriptors code_descriptors = {
...
         .bcdUSB             = USB_BCD_V20,
-        .bDeviceClass       = 0, // Class and protocol defined per interface
+        /* Class defined per interface */
+        .bDeviceClass       = 0, 
         .bDeviceSubClass    = 0,
+        /* Protocol defined per interface */
         .bDeviceProtocol    = 0,
-        .bMaxPacketSize0    = 64, // kB
+        /* packet size is in bytes */
+        .bMaxPacketSize0    = 64,
```
Comment (mithro) [inline audio/descriptors.c]: Why are you using 64 byte max packet size - can you add a comment here explaining why...

File: audio/descriptors.c [hunk 12]
```diff
@@ -26,10 +28,13 @@ __code __at(DSCR_AREA) struct usb_descriptors code_descriptors = {
         .bLength            = USB_DT_DEVICE_SIZE,
         .bDescriptorType    = USB_DT_DEVICE,
         .bcdUSB             = USB_BCD_V20,
-        .bDeviceClass       = 0, // Class and protocol defined per interface
+        /* Class defined per interface */
```
Comment (mithro) [inline audio/descriptors.c]: Why is the device/subclass zero? I think you want "Unused, the class is instead defined in the interface" or something similar...

File: audio/descriptors.c [hunk 13]
```diff
@@ -42,81 +47,235 @@ __code __at(DSCR_AREA) struct usb_descriptors code_descriptors = {
...
             .bLength                = USB_DT_CONFIG_SIZE,
             .bDescriptorType        = USB_DT_CONFIG,
             .wTotalLength           = sizeof(descriptors.highspeed),
-            .bNumInterfaces         = 1,
+            .bNumInterfaces         = 2,
             .bConfigurationValue    = 1,
             .iConfiguration         = 0,
             .bmAttributes           = USB_CONFIG_ATT_ONE,
-            .bMaxPower              = 0x32, // * 2 mA
+            /* bMaxPower has a resolution of 2mA */
```
Comment (mithro) [inline audio/descriptors.c:75]: Is this 0x32 == 2 * 0x32 == 50 * 2 == 100mA?

Probably shouldn't use hex here.... ```c
.bMaxPower = 100 / 2, /* bMaxPower is in 2mA increments */
```
Seems clearer?

File: audio/descriptors.c [hunk 14]
```diff
@@ -42,81 +47,235 @@ __code __at(DSCR_AREA) struct usb_descriptors code_descriptors = {
...
+            .classspec = {
+                .bLength            = UAC_DT_AC_HEADER_SIZE(1),
+                .bDescriptorType    = USB_DT_CS_INTERFACE,
+                .bDescriptorSubtype = UAC_MS_HEADER,
+                .bcdADC             = UAC_BCD_V10,
+                .wTotalLength       =
+                    sizeof(descriptors.highspeed.control.classspec) +
+                    sizeof(descriptors.highspeed.input) +
+                    sizeof(descriptors.highspeed.output),
+                /* Number of streaming interfaces */
```
Comment (mithro) [inline audio/descriptors.c]: Streaming interfaces? Audio data stream interfaces?

File: audio/descriptors.c [hunk 15]
```diff
@@ -42,81 +47,235 @@ __code __at(DSCR_AREA) struct usb_descriptors code_descriptors = {
...
+        .control = {
+            .standard = {
+                .bLength            = USB_DT_INTERFACE_SIZE,
+                .bDescriptorType    = USB_DT_INTERFACE,
+                .bInterfaceNumber   = 0,
+                .bAlternateSetting  = 0,
+                .bNumEndpoints      = 0,
+                .bInterfaceClass    = USB_CLASS_AUDIO,
+                .bInterfaceSubClass = USB_SUBCLASS_AUDIOCONTROL,
+                /* Must be zero */
```
Comment (mithro) [inline audio/descriptors.c]: Why?

File: audio/descriptors.c [hunk 16]
```diff
@@ -1,22 +1,24 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+
+/** \file descriptors.c
+ * Describes the device according to Section 4 of the USB Audio Spec
```
Comment (mithro) [inline audio/descriptors.c]: What type of device does it describe and what features does it have?

File: audio/descriptors.c [hunk 17]
```diff
@@ -42,81 +47,235 @@ __code __at(DSCR_AREA) struct usb_descriptors code_descriptors = {
...
+            /* 8000Hz, little endian */
+            .tSamFreq[0]        = { 0x40, 0x1F, 0x00 },
         },
         .endpoints = {
             {
                 .bLength            = USB_DT_ENDPOINT_AUDIO_SIZE,
                 .bDescriptorType    = USB_DT_ENDPOINT,
-                .bEndpointAddress   = USB_ENDPOINT_NUMBER(0x2) | USB_DIR_OUT,
-                .bmAttributes       = 0x1, // Isynchronous endpoint
+                .bEndpointAddress   = USB_ENDPOINT_NUMBER(8) | USB_DIR_IN,
```
Comment (mithro) [inline audio/descriptors.c]: Why did we use endpoint 8?

File: audio/descriptors.c [hunk 18]
```diff
@@ -42,81 +47,235 @@ __code __at(DSCR_AREA) struct usb_descriptors code_descriptors = {
...
+                .bmAttributes       = (USB_ENDPOINT_XFER_ISOC | USB_ENDPOINT_SYNC_NONE),
                 .wMaxPacketSize     = 512,
-                .bInterval          = 1,
+                .bInterval          = 4,
                 .bRefresh           = 0,
                 .bSynchAddress      = 0,
             },
         },
+        .isoendpoint = {
```
Comment (mithro) [inline audio/descriptors.c:190]: Why is the `isoendpoint` unused? Can you add a comment here about why?

File: audio/firmware.c [hunk 19]
```diff
@@ -1,201 +1,129 @@
...
+     * reset to default
+     * Invalid
+     * Type: Bulk
+     * Direction: out
+     * Buffer: Double
+     */
+    EP2CFG = 0x7F;
+    SYNCDELAY; EP4CFG = 0x7F;
+    SYNCDELAY; EP6CFG = 0x7F;
+    SYNCDELAY; EP8CFG = 0x7F;
```
Comment (mithro) [inline audio/firmware.c:77]: Should there be a "reset endpoints" type function in fx2lib?

File: audio/firmware.c [hunk 20]
```diff
@@ -1,201 +1,129 @@
...
-                bytes = MAKEWORD(EP2BCH,EP2BCL);
-
-                for (i=0;i<bytes;++i) EP6FIFOBUF[i] = EP2FIFOBUF[i];
-
-                // can copy whole string w/ autoptr instead.
-                // or copy directly from one buf to another
-
-                // ARM ep6 out
-                EP6BCH=MSB(bytes);
+        /* ISO endpoint config type is 01 in the enpoint configuration buffer */
```
Comment (mithro) [inline audio/firmware.c:93]: Why was the iso endpoint structure empty in the descriptors?

**Review by mithro — COMMENTED** [review]
Looking really good!

Just a couple of small changes / fixes.

File: audio/Makefile (hunk 1 above)
Comment (paddatrapper) [reply audio/Makefile]: Yup. This was the result of our debug session a while ago

File: audio/Makefile (hunk 2 above)
Comment (paddatrapper) [reply audio/Makefile:46]: Serial stuff for `printf(...)`

File: audio/audiodata.c (hunk 4 above)
Comment (paddatrapper) [reply audio/audiodata.c]: Following the video firmware

File: audio/debug.c (hunk 10 above)
Comment (paddatrapper) [reply audio/debug.c:23]: FX2 yes, I don't know of a pin on either the opsis or atlys I can pull for serial communication. Didn't manage to port serial over

File: audio/descriptors.c (hunk 16 above)
Comment (paddatrapper) [reply audio/descriptors.c]: An audio device. Section 4 is simple the descriptors section

File: audio/descriptors.c (hunk 12 above)
Comment (paddatrapper) [reply audio/descriptors.c]: 0 indicates the class is defined per interface

File: audio/descriptors.c (hunk 11 above)
Comment (paddatrapper) [reply audio/descriptors.c]: Don't know of a technical reason, except that is what the logitech reference used... This is just ep0 though

File: audio/descriptors.c (hunk 15 above)
Comment (paddatrapper) [reply audio/descriptors.c]: Because the spec says so

File: audio/descriptors.c (hunk 17 above)
Comment (paddatrapper) [reply audio/descriptors.c]: It's unused in the video firmware so merging the two is easier. Makes this work on the opsis

File: audio/firmware.c (hunk 19 above)
Comment (paddatrapper) [reply audio/firmware.c:77]: This isn't resetting, but disabling. The TRM says to leave the other bits as their defaults, so this is default, with the VALID bit unset

File: audio/descriptors.c (hunk 18 above)
Comment (paddatrapper) [reply audio/descriptors.c:190]: It is used...? Describes the ISO streaming

File: audio/firmware.c (hunk 20 above)
Comment (paddatrapper) [reply audio/firmware.c:93]: It isn't...?

File: audio/descriptors.c (hunk 12 above)
Comment (mithro) [reply audio/descriptors.c]: See, I had forgotten that and the comment didn't tell me that...

File: audio/descriptors.c (hunk 11 above)
Comment (mithro) [reply audio/descriptors.c]: Put a comment `/* FIXME: Unsure why we are using 64 bytes, copied from a logitech XYZ device. */`

File: audio/descriptors.c (hunk 15 above)
Comment (mithro) [reply audio/descriptors.c]: Spec xyz requires this to be zero....

File: audio/descriptors.c (hunk 17 above)
Comment (mithro) [reply audio/descriptors.c]: Comment would be useful :-P

File: audio/descriptors.c (hunk 18 above)
Comment (mithro) [reply audio/descriptors.c:190]: The `/* Unused */` comment in the middle confused me....

File: audio/firmware.c (hunk 19 above)
Comment (mithro) [reply audio/firmware.c:77]: "Restore all endpoints to their default settings as described by the TRM at XYZ."

File: audio/debug.c (hunk 8 above)
Comment (paddatrapper) [reply audio/debug.c:1]: Probably. Needs to have support for Opsis and Atlys though

File: audio/descriptors.c (hunk 11 above)
Comment (paddatrapper) [reply audio/descriptors.c]: Found it - it is the maximum size of a control packet, which is what EP0 handles. Added comment

File: audio/audiodata.c (hunk 5 above)
Comment (paddatrapper) [reply audio/audiodata.c:57]: Fixed

File: audio/audiodata.c (hunk 7 above)
Comment (paddatrapper) [reply audio/audiodata.c:102]: EP8, Fixed

File: audio/descriptors.c (hunk 12 above)
Comment (paddatrapper) [reply audio/descriptors.c]: Fixed

File: audio/descriptors.c (hunk 15 above)
Comment (paddatrapper) [reply audio/descriptors.c]: Fixed

File: audio/descriptors.c (hunk 18 above)
Comment (paddatrapper) [reply audio/descriptors.c:190]: Clarified

File: audio/firmware.c (hunk 19 above)
Comment (paddatrapper) [reply audio/firmware.c:77]: Fixed

**Comment by paddatrapper** [comment]
Travis and `version_data.h` again...

File: audio/Makefile [hunk 21]
```diff
@@ -38,10 +38,11 @@ LINUXHEADERSDIR := ../third_party/fx2lib-linux-headers
 COMMON_DIR := ../common
 INCS := -I$(LINUXHEADERSDIR)
 
-CFLAGS := -DDSCR_AREA=0x3e00 --code-size 0x3c00 --xram-size 0x0200
+# Descriptors require more space in memory
```
Comment (mithro) [inline audio/Makefile]: Maybe? `audio descriptors require more space, so using 3d00 rather than the normal 0x3c00`

File: audio/audiodata.c (hunk 4 above)
Comment (mithro) [reply audio/audiodata.c]: Why does the video firmware do that? :-P

Comment (paddatrapper) [reply audio/audiodata.c]: Looking at the descriptors, we only have one configuration - USB audio device for audio firmware and USB video device for the video firmware. So we can only work with that config. Reference: http://www.beyondlogic.org/usbnutshell/usb5.shtml#ConfigurationDescriptors

I'll update the comment

File: audio/firmware.c [hunk 22]
```diff
@@ -1,201 +1,130 @@
...
-    // arm ep2
-    EP2BCL = 0x80; // write once
-    SYNCDELAY;
-    EP2BCL = 0x80; // do it again
-
-    // make it so we enumerate
-    EA=1; // global interrupt enable 
-    printf ( "Done initializing stuff\n" );
-
+    /**
```
Comment (mithro) [inline audio/firmware.c:65]: You have used this construct multiple times, we should add it to fx2lib?

File: audio/reference/reference_descriptors [hunk 23]
```diff
@@ -1,783 +0,0 @@
-Bus 001 Device 058: ID 046d:082b Logitech, Inc. Webcam C170
```
Comment (mithro) [inline audio/reference/reference_descriptors:1]: Did you remove these files deliberately?

File: audio/firmware.c (hunk 22 above)
Comment (paddatrapper) [reply audio/firmware.c:65]: Probably a macro for `DISABLEEP` which sets the config for that to `0x7F` would help

File: audio/reference/reference_descriptors (hunk 23 above)
Comment (paddatrapper) [reply audio/reference/reference_descriptors:1]: Yes, as they are not much use past the initial setup of the descriptors

File: audio/descriptors.c (hunk 13 above)
Comment (paddatrapper) [reply audio/descriptors.c:75]: Fixed

File: audio/Makefile (hunk 21 above)
Comment (paddatrapper) [reply audio/Makefile]: Fixed

File: audio/Makefile (hunk 1 above)
Comment (paddatrapper) [reply audio/Makefile]: Fixed

File: audio/TODO (hunk 3 above)
Comment (paddatrapper) [reply audio/TODO]: Fixed

File: audio/audiodata.c (hunk 4 above)
Comment (paddatrapper) [reply audio/audiodata.c]: Fixed

File: audio/audiodata.c (hunk 6 above)
Comment (paddatrapper) [reply audio/audiodata.c]: Fixed

File: audio/descriptors.c (hunk 14 above)
Comment (paddatrapper) [reply audio/descriptors.c]: Fixed

File: audio/descriptors.c (hunk 17 above)
Comment (paddatrapper) [reply audio/descriptors.c]: Fixed

**Pull request merged**